act/qc/qcfilter.py

```diff
@@ -1127,6 +1146,125 @@ def datafilter(
+        if len(ancillary_variables) < 2:
+            raise ValueError('Data variable must have more than one ancillary variable.')
+
+        if qc_var_names is None:
+            qc_var_names = ancillary_variables
+
+        qc_var_name = 'qc_' + var_name + '_dummy'
+        if qc_var_name in self._ds.variables:
+            # Clear the dummy variable if it already exists
+            warnings.warn(
+                f'Dummy quality control variable {qc_var_name} already exists. It will be cleared and recreated.'
+            )
+            del self._ds[qc_var_name]
+
+        # Create the quality control variable with same dimensions as data variable.
+        qc_data = np.zeros(self._ds[var_name].shape, dtype=bool)
+        for qc_var in qc_var_names:
+            flag_value = False
+            flag_values = None
+            flag_masks = None
+            flag_assessments = None
+            try:
+                flag_assessments = self._ds[qc_var].attrs['flag_assessments']
+                flag_masks = self._ds[qc_var].attrs['flag_masks']
+            except KeyError:
+                pass
+
+            try:
+                flag_values = self._ds[qc_var].attrs['flag_values']
+                flag_value = True
+            except KeyError:
+                pass
+
+            test_numbers = []
+            if rm_tests is not None:
+                if isinstance(rm_tests, (int, float, str)):
+                    rm_tests = [int(rm_tests)]
+                test_numbers.extend(rm_tests)
+
+            if rm_assessments is not None:
+                if isinstance(rm_assessments, str):
+                    rm_assessments = [rm_assessments]
+
+                if flag_masks is not None:
+                    test_nums = [parse_bit(mask)[0] for mask in flag_masks]
+
+                if flag_values is not None:
+                    test_nums = flag_values
+
+                rm_assessments = [x.lower() for x in rm_assessments]
+                if flag_assessments is not None:
+                    for ii, assessment in enumerate(flag_assessments):
+                        if assessment.lower() in rm_assessments:
+                            test_numbers.append(test_nums[ii])
+
+            # Make the list of test numbers to mask unique
+            test_numbers = list(set(test_numbers))
+            for test in test_numbers:
+                qc_test_mask = self._ds.qcfilter.get_qc_test_mask(
+                    var_name, test, qc_var_name=qc_var, flag_value=flag_value
+                )
+                qc_data = qc_data | qc_test_mask
+            # If data was orginally stored as Dask array return values to Dataset as Dask array
+            # else set as Numpy array.
+        try:
+            self._ds[qc_var_name] = (self._ds[var_name].dims, dask.array.from_array(qc_data))
+        except AttributeError:
+            self._ds[qc_var_name] = (self._ds[var_name].dims, qc_data)
+
+        # Add attributes to the quality control variable.
```

Comment:
Should all the attributes be hard coded or should we be pulling this from existing?  Especially the flag_meanings and flag_assessments?